### PR TITLE
Allow exchanging forbidden jutsu/scroll

### PR DESCRIPTION
**Pull request by HitoriWoB**

### CODE REVIEW

File: classes/forbidden_shop/ForbiddenShopManager.php [hunk 1]
```diff
@@ -36,8 +48,23 @@ public function buyForbiddenJutsu($jutsu_id): string {
...
+                }
+                else {
+                    unset($this->player->jutsu_scrolls[$jutsu_scroll->id]);
+                    $remove_text = "You have exchanged {$jutsu_scroll->name} scroll ";
+                }
+                $this->player->jutsu_scrolls[$jutsu_id] = $jutsu;
+                $this->player->updateInventory();
+                return $remove_text . "in exchange for {$jutsu->name}!";
+            }
+            // Does not have scroll or forbidden jutsu/scroll to exchange
```
[the remark this answers is not in the document]
Comment (HitoriWoB) [reply classes/forbidden_shop/ForbiddenShopManager.php:67]: Refactoring this now.

File: classes/chat/ChatManager.php [hunk 2]
```diff
@@ -162,7 +162,7 @@ private function fetchPosts(?int $starting_post_id = null, int $max_posts = self
             $post->message = nl2br($this->system->html_parse($post->message, false, true));
 
             // Handle Mention
-            $pattern = "/@([^ \n\s!?.<>:@\[\]()]+)(?=[^A-Za-z0-9_]|$)/";
+            $pattern = "/@([^\n\s]+[a-zA-Z0-9_-]|$)/";
```
[the remark this answers is not in the document]
Comment (HitoriWoB) [reply classes/chat/ChatManager.php:165]: This fixes crashing the chat for things like:
@lsmjudoka/
@lsmjudoka\
@lsmjudoka*

It's also more resilient:
@lsm/judoka =>  [tag]@lsm[/tag] /judoka

[the remark this answers is not in the document]
Comment (HitoriWoB) [reply classes/chat/ChatManager.php:165]: image
Proof of concept

File: classes/forbidden_shop/ForbiddenShopManager.php [hunk 3]
```diff
@@ -20,6 +20,18 @@ public function buyForbiddenJutsu($jutsu_id): string {
...
+        $learned_jutsu = false;
+        $jutsu_scroll = false;
+        foreach($forbidden_jutsu as $fj_id => $fj) {
+            if($this->player->hasJutsu($fj_id)) {
+                $learned_jutsu = $fj;
+            }
+            if(isset($this->player->jutsu_scrolls[$fj_id])) {
+                $jutsu_scroll = $fj;
+            }
+        }
```
Comment (levimeahan) [inline classes/forbidden_shop/ForbiddenShopManager.php:33]: Is this part necessary now?

**Review by levimeahan — APPROVED** [review]

File: classes/forbidden_shop/ForbiddenShopManager.php (hunk 3 above)
Comment (HitoriWoB) [reply classes/forbidden_shop/ForbiddenShopManager.php:33]: It's not, I missed that part in removal

**Pull request merged**